### PR TITLE
Added instr.sched options to `tune_gemm.py`

### PR DESCRIPTION
**Pull request by ravil-mobile**

### CODE REVIEW

**Comment by xiaohuguo2023** [comment]
Can you add some description in the README file ?

**Comment by ravil-mobile** [comment]
> Do we know which gemm sizes can potentially benefit from setting `instruction_sched_variant` to iglp0 or ckv3?

No, at this moment it is difficult to say

**Review by sjw36 — CHANGES_REQUESTED** [review]
Looks good but need to wait for features in TIP. Blocking for now.

**Comment by ravil-mobile** [comment]
> Looks good but need to wait for features in TIP. Blocking for now.

Sorry. I didn't get what you mean

**Comment by zhanglx13** [comment]
> > Do we know which gemm sizes can potentially benefit from setting `instruction_sched_variant` to iglp0 or ckv3?
> 
> No, at this moment it is difficult to say

If this is the case, I'd say we don't want to add more tuning parameters. 
If we know it works for "some" kernels for sure, we should use them by some simple heuristics.
If we don't know when they help, we should figure it out first.

**Review by zhanglx13 — CHANGES_REQUESTED** [review]
See my comments

**Comment by ravil-mobile** [comment]
> > > Do we know which gemm sizes can potentially benefit from setting `instruction_sched_variant` to iglp0 or ckv3?
> > 
> > 
> > No, at this moment it is difficult to say
> 
> If this is the case, I'd say we don't want to add more tuning parameters. If we know it works for "some" kernels for sure, we should use them by some simple heuristics. If we don't know when they help, we should figure it out first.

Ok. Let me make `sched_variants = ["\"default\""]`. This means that only *no scheduling* variant is going to work. Tell me whether you are ok with it. At least we should push the adapted infrastructure to `main_perf`

File: python/perf-kernels/tools/tune_gemm/tune_gemm.py [hunk 1]
```diff
@@ -649,11 +653,11 @@ def main():
         formatted_tflops = format_output(tri_tflops)
         minTime = format_output(minTime)
         if not run_bench:
-            print(f'TFLOPS: {formatted_tflops} time(us): {minTime}', end=" ", flush=True)
+            print(f'\nTFLOPS: {formatted_tflops}; time(us): {minTime}', end=" ", flush=True)
 
         bestConfig_compact_str = gen_configStr(bestConfig)
         if not run_bench:
-            print(f'best_config: {bestConfig_compact_str}', end=" ", flush=True)
+            print(f'\nbest_config: {bestConfig_compact_str}', end=" ", flush=True)
```
Comment (zhanglx13) [inline python/perf-kernels/tools/tune_gemm/tune_gemm.py:663]: Do you have an example output after adding '\n'?

File: python/perf-kernels/tools/tune_gemm/tune_gemm.py [hunk 2]
```diff
@@ -72,13 +73,15 @@ def get_full_tuning_space():
...
                                 for waves_per_eu in waves_per_eu_range:
                                     for matrix_instr_nonkdim in matrix_instr_nonkdim_range:
-                                        for kpack in kpack_range:
-                                            configs.append({
-                                                'BLOCK_SIZE_M': block_m, 'BLOCK_SIZE_N': block_n, 'BLOCK_SIZE_K':
-                                                block_k, 'GROUP_SIZE_M': group_m, 'SPLIT_K': split_k, 'num_warps':
-                                                num_warps, 'num_stages': num_stages, 'waves_per_eu': waves_per_eu,
-                                                'matrix_instr_nonkdim': matrix_instr_nonkdim, 'kpack': kpack
-                                            })
+                                        for sched_variant in sched_variants:
```
Comment (zhanglx13) [inline python/perf-kernels/tools/tune_gemm/tune_gemm.py]: Since you are here, can you replace the nested for loops with itertools.product?

**Comment by zhanglx13** [comment]
> Ok. Let me make sched_variants = ["\"default\""]. This means that only no scheduling variant is going to work. Tell me whether you are ok with it. At least we should push the adapted infrastructure to main_perf

Yes, I'm ok with it

File: python/perf-kernels/tools/tune_gemm/tune_gemm.py (hunk 1 above)
Comment (ravil-mobile) [reply python/perf-kernels/tools/tune_gemm/tune_gemm.py:663]: Yes, sure.

```bash
> ./tune_gemm.py --gemm_size_file ~/tuning/input.yaml --gpu_ids 3,4,5 --jobs 32 --o ~/tuning/output.yaml
Tuning 1 gemm sizes starts at: 2024-10-29 14:26:32.618604
SIZE: 4864 8192 4160 TN nConfigs: 720 
TFLOPS: 516.47; time(us): 641.89 
best_config: BM128_BN128_BK64_GM8_SK1_nW4_nS2_EU0_kP2_mfma16_schedDEFAULT 
>>> Elapsed time: 0:04:11.238153 = 0:00:20.441773 (compile) + 0:03:49.947198 (profile) + 0:00:00.681876 (post processing)
Tuning ends at: 2024-10-29 14:30:44.031012
Total tuning time (h:m:s): 0:04:11.412408
```

File: python/perf-kernels/tools/tune_gemm/tune_gemm.py (hunk 2 above)
Comment (ravil-mobile) [reply python/perf-kernels/tools/tune_gemm/tune_gemm.py]: Yes, sure. Done!

File: python/perf-kernels/tools/tune_gemm/utils/file_generator.py [hunk 3]
```diff
@@ -112,6 +113,7 @@ def matmul_{configStr}(M, N, K, am, ak, bk, bn, cm, cn, biasn):
         EVEN_K = {EVEN_K},
         GRID_MN = grid_mn,
         NUM_XCDS = {num_xcds},
+        instruction_sched_variant = {sched_variant},
```
Comment (zhanglx13) [inline python/perf-kernels/tools/tune_gemm/utils/file_generator.py]: Missing quotes `' '`

File: python/perf-kernels/tools/tune_gemm/utils/file_generator.py [hunk 4]
```diff
@@ -145,7 +147,8 @@ def matmul_{configStr}(a, b, c, bias, M, N, K, am, ak, bk, bn, cm, cn, biasn):
         BIAS = {use_bias},
         EVEN_K = {EVEN_K},
         GRID_MN = grid[0],
-        NUM_XCDS = {num_xcds}
+        NUM_XCDS = {num_xcds},
+        instruction_sched_variant = {sched_variant},
```
Comment (zhanglx13) [inline python/perf-kernels/tools/tune_gemm/utils/file_generator.py]: Missing quote `' '`

File: python/perf-kernels/tools/tune_gemm/utils/file_generator.py [hunk 5]
```diff
@@ -21,15 +21,16 @@ def read_config(config):
...
 
 
 def gen_configStr(config):
-    block_m, block_n, block_k, group_m, split_k, num_warps, num_stages, waves_per_eu, mfmaInstrSize, kpack = read_config(
+    block_m, block_n, block_k, group_m, split_k, num_warps, num_stages, waves_per_eu, mfmaInstrSize, kpack, sched_variant = read_config(
         config)
 
     ## {M}_{N}_{K} is removed since the same kernel can be used for differen gemm sizes
-    configStr = f"BM{block_m}_BN{block_n}_BK{block_k}_GM{group_m}_SK{split_k}_nW{num_warps}_nS{num_stages}_EU{waves_per_eu}_kP{kpack}_mfma{mfmaInstrSize}"
+    configStr = f"BM{block_m}_BN{block_n}_BK{block_k}_GM{group_m}_SK{split_k}_nW{num_warps}_nS{num_stages}_EU{waves_per_eu}_kP{kpack}_mfma{mfmaInstrSize}_sched{sched_variant[1:-1].upper()}"
```
Comment (zhanglx13) [inline python/perf-kernels/tools/tune_gemm/utils/file_generator.py]: Now we are using `local-prefetch`, but we cannot have `-` in kernel names. Can you also convert `-` into `_`?

**Review by zhanglx13 — CHANGES_REQUESTED** [review]
See my comments

File: python/perf-kernels/tools/tune_gemm/utils/file_generator.py (hunk 4 above)
Comment (ravil-mobile) [reply python/perf-kernels/tools/tune_gemm/utils/file_generator.py]: done

File: python/perf-kernels/tools/tune_gemm/utils/file_generator.py (hunk 5 above)
Comment (ravil-mobile) [reply python/perf-kernels/tools/tune_gemm/utils/file_generator.py]: done

File: python/perf-kernels/tools/tune_gemm/test_regression.py [hunk 6]
```diff
@@ -101,6 +101,7 @@ def teardown_class(self):
         },
     ], ids=lambda val: f"Config: {val}")
     def test_matmul_performance_regression(self, config, record_property):
+        config.setdefault('instruction_sched_variant', 'default')
```
Comment (sjw36) [inline python/perf-kernels/tools/tune_gemm/test_regression.py]: should be `none`?

**Review by sjw36 — APPROVED** [review]
Just one question. Otherwise LGTM.

File: python/perf-kernels/tools/tune_gemm/test_regression.py (hunk 6 above)
Comment (ravil-mobile) [reply python/perf-kernels/tools/tune_gemm/test_regression.py]: Thanks! I missed it somehow.

Comment (ravil-mobile) [reply python/perf-kernels/tools/tune_gemm/test_regression.py]: Done

**Review by zhanglx13 — APPROVED** [review]

**Pull request merged**